### ISSUE TITLE
Unexpected _run_once_on_dataset behavior if using a custom dataloader that implements __len__ method

### ISSUE THREAD

**Issue by AaamberW** [pytorch/ignite#2610]
## 🐛 Bug description

Context - I was calling engine.run(my_run_function), with a custom dataloader that implements `__len__` method.

What happened - Looks like when my run hits [_run_once_on_dataset](https://github.com/pytorch/ignite/blob/22223ba1e83569e66781ab7b98ca792b4d85bca1/ignite/engine/engine.py#L788), the expected behavior is to break out of while loop if dataloader raises **StopIteration**. Based on another [discussions](https://github.com/pytorch/ignite/issues/1094), we expect `self.state.epoch_length` to be None therefore hit break from [there](https://github.com/pytorch/ignite/blob/22223ba1e83569e66781ab7b98ca792b4d85bca1/ignite/engine/engine.py#L816). However during [previous steps](https://github.com/pytorch/ignite/blob/22223ba1e83569e66781ab7b98ca792b4d85bca1/ignite/engine/engine.py#L686), `self.state.epoch_length` would be set to a positive integer since `self.state.max_epochs`  is None for a new run(instead of reading from state_dict) and I have `__len__` method defined, even if I didn't provide `epoch_length` param to `run` function.

What I expect - Upon StopIteration, dataloader is exhausted for one epoch, exit `_run_once_on_dataset` without reinitializing dataloader and re-enter the while loop.

## Environment
- PyTorch Version (e.g., 1.4): 1.10
- Ignite Version (e.g., 0.3.0): 0.4.9
- OS (e.g., Linux): MacOS
- How you installed Ignite (`conda`, `pip`, source): pip
- Python version: 3.9
- Any other relevant information:

**Comment by vfdev-5** [comment]
Thanks for reporting @AaamberW 
Could you please provide a minimal code to reproduce the problem ? I'm a bit confused about what's exactly the problem... Thanks!

**Comment by AaamberW** [comment]
Hi @vfdev-5, here's a simplified version of my code 
`
class CustomDataLoader:
    def __init__():
        ...
        print('custom data loader init')

    def __len__(self):
        ...
    
    def __next__(self):
        if not data_exhausted:
            ...
            print (f"Loading batch {batch_id}")
        else:
            raise StopIteration
train_data = CustomDataLoader() 

def do_train(engine, batch):
    optimizer.zero_grad()
    batch = [x.to(device) for x in batch]
    model.train()
    losses = model(*batch)
    loss = losses.mean()
    loss.backward()
    optimizer.step()
    return {"loss": loss.item()}

trainer = Engine(do_train)
trainer.run(train_data, max_epochs=1)
    
`
I have 200 batches total. With max_epoches=1, I expect CustomDataLoader being initialized once but from the logs I got
`
custom data loader init
Loading batch 1
Loading batch 2
...
Loading batch 200
custom data loader init
Loading batch 1
Loading batch 2
...
`

I encountered it while upgrading from 0.2.1 to any version>= 0.3.0. I traced it to [this line ](https://github.com/pytorch/ignite/blob/22223ba1e83569e66781ab7b98ca792b4d85bca1/ignite/engine/engine.py#L832) being called after all 200 batches finish for 1 epoch, which doesn't look like the correct behavior. Happy to further clarify.

**Comment by vfdev-5** [comment]
@AaamberW thanks for the code snippet. I updated it to make it executable and tried to reproduce your behavior but engine worked as expected:
```python
class CustomDataLoader:
    def __init__(self, size=10):
        print('custom data loader init')
        self.i = 0
        self.size = size

    def __len__(self):
        return self.size
    
    def __next__(self):
        data_exhausted = self.i >= len(self)
        if not data_exhausted:
            print (f"Loading batch {self.i}")
            self.i += 1
            return 0.1 * self.i
        else:
            raise StopIteration

    def __iter__(self):
        return self    
            

train_data = CustomDataLoader() 

from ignite.engine import Engine

def do_train(engine, batch):
    return 0.0

trainer = Engine(do_train)
trainer.run(train_data, max_epochs=1)
```
Output:
```
custom data loader init
Loading batch 0
Loading batch 1
Loading batch 2
Loading batch 3
Loading batch 4
Loading batch 5
Loading batch 6
Loading batch 7
Loading batch 8
Loading batch 9

State:
	iteration: 10
	epoch: 1
	epoch_length: 10
	max_epochs: 1
	max_iters: <class 'NoneType'>
	output: 0.0
	batch: 1.0
	metrics: <class 'dict'>
	dataloader: <class '__main__.CustomDataLoader'>
	seed: <class 'NoneType'>
	times: <class 'dict'>
```

Could you please update my code snippet to reproduce your behaviour ?

**Comment by AaamberW** [comment]
Hi @vfdev-5, thanks for the script. Looks like if `self.state.epoch_length` is set, then it relies on[ this line ](https://github.com/pytorch/ignite/blob/22223ba1e83569e66781ab7b98ca792b4d85bca1/ignite/engine/engine.py#L849) to identify epoch completion rather than dataloader raising StopIteration. 

In my case, CustomDataLoader fetches one partition from N total partitions each iteration, and partitions come in varying batches sizes. 

`for par in partitions: #load multiple batches in one partition in memory
        for batch in par:
             yield batch
`
When CustomDataLoader finished loading the last partition, `iteration_counter` didn't reach `epoch_length`, therefore even though CustomDataLoader raised StopIteration my code still [hits reinit data loader ](https://github.com/pytorch/ignite/blob/22223ba1e83569e66781ab7b98ca792b4d85bca1/ignite/engine/engine.py#L832). 

To illustrate, imagine a CustomDataLoader trying to fetch 5 partitions of data that contains 50 batches in total, with the following schedule
`
#partition | p1 | p2 | p3 | p4 | p5
#batches  | 1   | 1   | 19  | 10  | 19
`
By p5 CustomDataLoader raised StopIteration since it finished loading all 50 batches, but iteration_counter was 1+1+19+10+1 by then therefore ignored the StopIteration and initialized CustomDataLoader again. Would it be possible to make StopIteration work again here?

**Comment by vfdev-5** [comment]
> Would it be possible to make StopIteration work for my case again here?

Yes, I'm trying to understand your use-case and figure out from my side what could be go wrong. RIght now, I have to find your CustomDataloader-like implementation that could help reproduce the issue.
